### PR TITLE
Move to Vue 3

### PR DESCRIPTION
**Pull request by dobromir-hristov**
## Description

Updates the library to support Vue 3.

Had to hack around Vuepress as it was giving errors with Vue CLI. 
Vue-server-renderer used in core tests is not working atm. (needs the new renderer for vue 3)
Adjusted ESLint and some deps to the root.
Added darkmode to test-project :)

### CODE REVIEW

File: .eslintrc.js [hunk 1]
```diff
@@ -1,19 +1,13 @@
...
-    browser: true,
-    es6: true,
     node: true
   },
   extends: [
-    // 'plugin:vue/essential', // is this required?
+    'plugin:vue/recommended',
     'standard'
   ],
-  globals: {},
```
Comment (dobromir-hristov) [inline .eslintrc.js]: These are part of the preset we extend

File: jest.base.js [hunk 2]
```diff
@@ -4,7 +4,7 @@ const dir = (p) => path.resolve(__dirname, p)
 module.exports = {
   rootDir: '.',
   transform: {
-    '^.+\\.js$': [dir('node_modules/babel-jest'), { cwd: __dirname }]
+    '^.+\\.js$': ['babel-jest', { cwd: __dirname }]
```
Comment (dobromir-hristov) [inline jest.base.js]: this is now hoisted to the top.

File: package.json [hunk 3]
```diff
@@ -3,20 +3,29 @@
...
   "private": true,
   "description": "Simple, lightweight model-based validation for Vue.js",
-  "workspaces": [
-    "packages/*"
-  ],
+  "workspaces": {
+    "packages": [
+      "packages/*"
+    ],
+    "nohoist": ["@vuelidate/docs/**"]
```
Comment (dobromir-hristov) [inline package.json]: This is causing all the headaches. I hope Vuepress gets an update soon.

File: packages/test-project/src/components/NestedA.vue [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+        >
+      </label>
+      <label>Min
+        <input
+          v-model.number="min"
+          type="number"
+        >
+      </label>
+    </div>
+    <NestedB
```
Comment (dobromir-hristov) [inline packages/test-project/src/components/NestedA.vue]: Here we try out passing params as props

File: packages/test-project/src/components/NestedB.vue [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+
+<script>
+import { ref, toRefs } from 'vue'
+import useVuelidate from '@vuelidate/core/src'
+import { required, maxValue, minValue } from '@vuelidate/validators/src/withMessages'
+
+export default {
+  name: 'NestedB',
+  setup (props) {
+    const { max, min } = toRefs(props)
```
Comment (dobromir-hristov) [inline packages/test-project/src/components/NestedB.vue]: This is the only way I managed to make this work.

File: packages/test-project/src/components/NestedValidations.vue [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+
+export default {
+  components: { NestedA },
+  setup () {
+    const numberX = ref(0)
+    const conditionalParam = ref('')
+
+    const validations = computed(() => {
+      const v = { numberX: { required, minValue: minValue(3) } }
+      if (numberX.value > 5) {
```
Comment (dobromir-hristov) [inline packages/test-project/src/components/NestedValidations.vue]: This never updates, unless I print the `validations` variable in the template.

File: packages/test-project/vue.config.js [hunk 7]
```diff
@@ -0,0 +1,3 @@
+module.exports = {
+  transpileDependencies: ['@vuelidate/core', '@vuelidate/validators']
```
Comment (dobromir-hristov) [inline packages/test-project/vue.config.js]: Important for sub-packages to be transpiled in test-project.

File: packages/vuelidate/src/core.js [hunk 8]
```diff
@@ -233,34 +236,35 @@ function createValidationResults (rules, state, key, parentKey) {
...
     result.$invalid.value && $dirty.value
   )
 
   result.$errors = computed(() => ruleKeys
-    .filter(ruleKey => unwrap(result[ruleKey]).$invalid)
+    .filter(ruleKey => result[ruleKey].$invalid.value)
     .map(ruleKey => {
       const res = result[ruleKey]
-      return {
+      return reactive({
```
Comment (dobromir-hristov) [inline packages/vuelidate/src/core.js]: Unwraps the errors, so we no longer have issues with refs in templates.

File: packages/vuelidate/src/core.js [hunk 9]
```diff
@@ -437,13 +442,13 @@ export function setValidations ({ validations, state, key, parentKey, childResul
    * If we have no `key`, this is the top level state
    * We dont need `$model` there.
    */
-  let $model = key ? '' : computed({
+  let $model = key ? computed({
```
Comment (dobromir-hristov) [inline packages/vuelidate/src/core.js]: This was broken. Now it works per old Vuelidate spec, all tree levels have a $model. The top one is just null in these cases.

File: packages/test-project/src/components/NestedB.vue (hunk 5 above)
Comment (shentao) [reply packages/test-project/src/components/NestedB.vue]: I think this is the recommended way.

File: packages/vuelidate/src/index.js [hunk 10]
```diff
@@ -40,16 +39,14 @@ export default function useVuelidate (validationsArg, state, registerAs) {
...
-      }
-    }
-  })
+  if (registerAs && childResultsKeys.value.length) {
+    return reactive({
+      ...validationResults,
+      ...childResults
+    })
+  } else {
+    return validationResults
```
Comment (shentao) [inline packages/vuelidate/src/index.js]: Why do we use `reactive` on the previous return statement but not here?

Comment (dobromir-hristov) [reply packages/vuelidate/src/index.js]: I thought it would make things more consistent, as the `else` returns a reactive object too.

Comment (shentao) [reply packages/vuelidate/src/index.js]: oh ok! We reactive it somewhere else. Makes sense!

**Pull request merged**